arch/sim/src/sim/up_timer.c

```diff
@@ -0,0 +1,169 @@
+
+#include "../../../../include/nuttx/config.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+typedef void (*sched_timer_callback_t)(void);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static sched_timer_callback_t g_sched_process_timer_cb;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: host_signal_handler
+ *
+ * Description:
+ *  The signal handler is called periodically and is used to deliver TICK
+ *  events to the OS.
+ *
+ ****************************************************************************/
+static void host_signal_handler(int sig, siginfo_t *si, void *old_ucontext)
+{
+  g_sched_process_timer_cb();
+}
+
+/****************************************************************************
+ * Name: host_setup_timer
+ *
+ * Description:
+ *  Set up a timer to send periodic signals.
+ *
+ ****************************************************************************/
+
+static int host_setup_timer(void)
+{
+  int ret;
+  struct itimerval it;
+
+  it.it_interval.tv_sec  = 0;
+  it.it_interval.tv_usec = CONFIG_SIM_PREEMPTIBLE_TASK_DURATION_MS * 1000;
+  it.it_value            = it.it_interval;
+
+  ret = setitimer(ITIMER_REAL, &it, NULL);
+  if (ret < 0)
+    {
+      fprintf(stderr, "%d settimer\n", ret);
+      return ret;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: host_setup_signals
+ *
+ * Description:
+ *  Set up a signal to deliver periodic TICK events.
+ *
+ ****************************************************************************/
+
+static int host_setup_signals(void (*action)(int, siginfo_t *, void *))
+{
+  struct sigaction act;
+  int ret;
+  sigset_t set;
+
+  act.sa_sigaction = action;
+  sigemptyset(&act.sa_mask);
+  act.sa_flags = SA_SIGINFO;
+
+  sigemptyset(&set);
+  sigaddset(&set, SIGALRM);
+
+  if ((ret = sigaction(SIGALRM, &act, NULL)) != 0)
+    {
+      fprintf(stderr, "%d signal handler", ret);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_init_timer
+ *
+ * Description:
+ *  Creates a periodic timer and sets up a signal handler on the host.
+ *
+ ****************************************************************************/
+
+int up_init_timer(void (*nxsched_process_timer)(void))
```

Comment:
I think this logic is better integrate into up_oneshot.c. The timer interrupt generation is very easy, the hardest part is how to call nxsched_process_timer without synchronization issue.